### ISSUE TITLE
Duplicate handlers in case of ContravariantRegistrationSource

### ISSUE THREAD

**Issue by rsivanov** [rebus-org/Rebus.Autofac#8]
Hello!

There is a problem with the current implementation of AutofacHandlerActivator. In case an application uses ContravariantRegistrationSource the handlers will be duplicated due to this code:
   var typesToResolve = messageType.GetBaseTypes()
                    .Concat(new[] {messageType})

ContravariantRegistrationSource already returns handlers for base types automatically, so you just add them one more time.

The possible solution is to add Distinct before returning the list of handlers:
return types.SelectMany(type => (IEnumerable<IHandleMessages<TMessage>>) lifetimeScope.Resolve(type)).Distinct();

**Comment by mookid8000** [comment]
Ah, I didn't know Autofac could do that....

I would very much love to NOT have to call `Resolve` on the container multiple times, and then leave it to the container to carry out the compatible types lookup.

Wouldn't it be nicer if Rebus simply called `Resolve` on the container, and then the user could enable polymorphic dispatch by adding the `ContravariantRegistrationSource`?

**Comment by rsivanov** [comment]
Yes, it would be a better solution, but incompatible with the applications that rely on your current method and don't use ContravariantRegistrationSource. So it will be a breaking change.

**Comment by mookid8000** [comment]
>  So it will be a breaking change.

yeah, well – I don't think that many people are actually using the polymorphic capabilities of Rebus.... and I bet it would yield much better handler resolution performance if it didn't have to perform (at least one, often times more) unnecessary handler lookups for each message.

IOW I think it would be worth it 😁

**Comment by mookid8000** [comment]
Rebus.Autofac 6.0.0-b01 now relies on Autofac to provide contravariant handler lookup.

This means that polymorphic dispatch must be enabled via Autofac like this:
```csharp
containerBuilder.RegisterSource(new ContravariantRegistrationSource());
```

This also means that handler lookup is probably much much faster 🏆 

Thanks for reporting this issue 😄

**Comment by dariogriffo** [comment]
So, there is no way of fixing this on 5.* versions?
I'm just having this issue, and the best way I have to hanlde this is <nasty>to rely on a flag in the IFailed handler</nasty>

**Comment by dariogriffo** [comment]
Well sorry, I will rephrase: does Rebus.Autofac needs to be 6.0*-beta yet after 1 year?
I have tested now the package and my problem was solved.
My scenario:
```cs
containerBuilder.RegisterSource(new ContravariantRegistrationSource());
```
is added, and I have a 
IHandleMessages<IFailed<ReportingExportCommand>>
that was being called twice, and sending 2 emails because of that.
I have upgraded the package and that seems to be working just fine, only 1 email sent.
Good work, I will continue testing my app during the day and confirm no other issues are created with the beta package.
